nilearn/connectivity/embedding.py

```diff
@@ -0,0 +1,394 @@
+        Maximal number of iterations.
+
+    tol : float, optional (default to 1e-7)
+        Tolerance.
+
+    Returns
+    -------
+    gmean : numpy.ndarray, shape (n_features, n_features)
+        Geometric mean of the matrices.
+    """
+    # Shape and symmetry positive definiteness checks
+    n_features = mats[0].shape[0]
+    for mat in mats:
+        _check_mat(mat, 'square')
+        if mat.shape[0] != n_features:
+            raise ValueError("Matrices are not of the same shape.")
+        _check_mat(mat, 'spd')
+
+    # Initialization
+    mats = np.array(mats)
+    if init is None:
+        gmean = np.mean(mats, axis=0)
+    else:
+        _check_mat(init, 'square')
+        if init.shape[0] != n_features:
+            raise ValueError("Initialization has not the correct shape.")
+        _check_mat(init, 'spd')
+        gmean = init
+
+    norm_old = np.inf
+    step = 1.
+
+    # Gradient descent
+    for n in range(max_iter):
+        # Computation of the gradient
+        vals_gmean, vecs_gmean = linalg.eigh(gmean)
+        gmean_inv_sqrt = _map_eig(np.sqrt, 1. / vals_gmean, vecs_gmean)
+        whitened_mats = [gmean_inv_sqrt.dot(mat).dot(gmean_inv_sqrt)
+                         for mat in mats]
+        logs = [_map_sym(np.log, w_mat) for w_mat in whitened_mats]
+        logs_mean = np.mean(logs, axis=0)  # Covariant derivative is
+                                           # - gmean.dot(logms_mean)
+        if np.any(np.isnan(logs_mean)):
+            raise FloatingPointError("Nan value after logarithm operation.")
+
+        norm = np.linalg.norm(logs_mean)  # Norm of the covariant derivative on
+                                          # the tangent space at point gmean
+
+        # Update of the minimizer
+        vals_log, vecs_log = linalg.eigh(logs_mean)
+        gmean_sqrt = _map_eig(np.sqrt, vals_gmean, vecs_gmean)
+        # Move along the geodesic
+        gmean = gmean_sqrt.dot(
+            _map_eig(np.exp, vals_log * step, vecs_log)).dot(gmean_sqrt)
+
+        # Update the norm and the step size
+        if norm < norm_old:
+            norm_old = norm
+        if norm > norm_old:
+            step = step / 2.
+            norm = norm_old
+        if tol is not None and norm / gmean.size < tol:
+            break
+    if tol is not None and norm / gmean.size >= tol:
+        warnings.warn("Maximum number of iterations {0} reached without "
+                      "getting to the requested tolerance level "
+                      "{1}.".format(max_iter, tol))
+
+    return gmean
+
+
+def sym_to_vec(sym, isometry=True):
+    """Return the flattened lower triangular part of an array.
+
+    Acts on the last two dimensions of the array if not 2-dimensional.
+
+    Parameters
+    ----------
+    sym : numpy.ndarray, shape (..., p, p)
+        Input array.
+
+    isometry : bool, optional (default to True)
+        If True, off diagonal terms of sym are multiplied by sqrt(2).
+
```

Comment:
Do we really need the option. In the current code base, the default is used systematically.